accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+
+2022.11.11
+  - Initial version, which is a copy of the views proposal, renaming 'view'
+    to 'inline', and adjusting the text accordingly. Also remove support for
+    primary constructors.
+
+
+## Summary
+
+This document specifies a language feature that we call "inline classes".
+
+The feature introduces _inline classes_, which is a new kind of class
+declared by a new `inline class` declaration. An inline class provides a
+replacement of the members available on instances of an existing type:
+when the static type of the instance is an inline type _V_, the available
+instance members are exactly the ones provided by _V_ (noting that
+there may also be some accessible and applicable extension members).
+
+In contrast, when the static type of an instance is not an inline type,
+it is (by soundness) always the run-time type of the instance or a
+supertype thereof. This means that the available instance members are
+the members of the run-time type of the instance, or a subset thereof
+(and there may also be some extension members).
+
+Hence, using a supertype as the static type allows us to see only a
+subset of the members. Using an inline type allows us to _replace_ the
+set of members, with subsetting as a special case.
+
+This functionality is entirely static. Invocation of an inline member is
+resolved at compile-time, based on the static type of the receiver.
+
+An inline class may be considered to be a zero-cost abstraction in the
+sense that it works similarly to a wrapper object that holds the
+wrapped object in a final instance variable. The inline class thus
+provides an interface which is chosen independently of the interface
+of the wrapped object, and it provides implementations of the members
+of the interface of the inline class, and those implementations can use
+the members of the wrapped object as needed.
+
+However, even though an inline class behaves like a wrapping, the wrapper
+object will never exist at run time, and a reference whose type is the
+inline class will actually refer directly to the underlying wrapped
+object. Every member access (e.g., an invocation of a method or a
+getter) on an expression whose static type is an inline type will invoke
+a member of the inline class (with some exceptions, as explained below),
+but this occurs because those member accesses are resolved statically,
+which means that the wrapper object is not actually needed.
+
+Given that there is no wrapper object, we will refer to the "wrapped"
+object as the _representation object_ of the inline class, or just the
+_representation_.
+
+Inside the inline class declaration, the keyword `this` is a reference
+to the representation whose static type is the enclosing inline
+class. A member access to a member of the enclosing inline class may
+rely on `this` being induced implicitly (for example, `foo()` means
+`this.foo()` if the inline class contains a method declaration named
+`foo`).
```

Comment:
, and there is no other `foo` in the lexical scope.

(If we have inheritance of inline class members, the `foo` of `this` might not be in the lexical scope, and another top-level declaration might be.)